src/ext/detective.lisp

```diff
@@ -0,0 +1,282 @@
+(defpackage :lem/detective
+  (:use :cl :lem)
+  (:export
+
+   :reference
+   :reference-name
+   :reference-point
+
+   :function-reference
+   :function-arguments
+
+   :package-reference
+   :package-use
+   :package-export
+
+   :class-reference
+   :class-parents
+   :class-attributes
+
+   :variable-reference
+   :variable-reference-value
+
+   :misc-reference
+   :misc-custom-type
+
+   :capture-regex
+   :make-capture-regex
+
+   :detective-search
+
+   :search-regex
+
+   :search-function-regex
+   :set-function-regex
+   :search-package-regex
+   :set-package-regex
+   :search-class-regex
+   :set-class-regex
+   :search-variable-regex
+   :set-variable-regex
+   :search-misc-regex
+   :set-misc-regex
+
+   :buffer-references
+
+   :search-references
+
+   :capture-references
+
+   :navigate-reference
+
+   :move-to-reference)
+  
+  #+sbcl
+  (:lock t))
+
+(in-package :lem/detective)
+
+(defclass reference ()
+  ((name :type string 
+         :initarg :reference-name
+         :reader reference-name)
+
+   (point :initarg :reference-point
+          :reader reference-point)))
+
+(defclass function-reference (reference)
+  ((arguments :type list
+              :initarg :function-arguments
+              :reader function-arguments)))
+
+(defclass package-reference (reference)
+  ((use :type list
+        :initarg :package-use
+        :reader package-use )
+   (export :type list
+           :initarg :package-export
+           :reader package-export)))
+
+(defclass class-reference (reference)
+  ((parents :type list
+            :initarg :class-parents
+            :reader class-parents)
+   (attributes :type list
+               :initarg :class-attributes
+               :reader class-attributes)))
+
+(defclass variable-reference (reference)
+  ((value :initarg :variable-reference-value
+          :reader variable-reference-value)))
+
+
+(defclass misc-reference (reference)
+  ((custom-type :initarg :misc-custom-type
+                :type string
+                :reader misc-custom-type)))
+
+(defstruct capture-regex regex function)
+
+(defclass detective-search () ())
```

Comment:
This name may be confusing as it seems to be covered with the symbol for `lem/language-mode:detective-search`.
Since it seems to be an abstract class, how about something like `<detective-search>`?

Reply:
Oh, I wasn't aware of that way of naming abstract classes, I changed it :+1: